Fixes #35661 - Handle ACS permissions on UI

sjha4 opened the pull request:
#### What are the changes introduced in this pull request?
Show/hide actions on ACS UI based on user permissions
#### Considerations taken when implementing this change?

#### What are the testing steps for this pull request?
Create a user with viewer role and refresh ACS UI after impersonating the user.
Make sure actions like refresh/delete/create/edit are hidden for the read-only user.
Try other combinations of ACS permissions and make sure actions are displayed/hidden accordingly.

ianballou commented:
It looks like the viewer for some reason doesn't see the Alternate Content Sources link on the side menu. Unless your branch here is out of date and doesn't have that change?

ianballou commented:
I tested the permissions and it looks fine! Only the permissions issue with actually seeing the Alternate  Content Sources link is there.

ianballou commented on app/views/katello/api/v2/alternate_content_sources/permissions.rabl:
```diff
@@ -0,0 +1,8 @@
+collection @alternate_content_sources
+
+if params.key?(:include_permissions)
```
`/katello/api/v2/alternate_content_sources?include_permissions=false` also causes the permissions to be returned.

ianballou replied:
Why is the create permission not included by default?

sjha4 replied:
> Why is the create permission not included by default?

It would be an additional meaningless(?) field for others consuming our API. The permission is only needed for the UI. 

Will update this: `if params.key?(:include_permissions)`

ianballou replied:
I meant more like why do you only need "can_create" and not "can_edit" or "can_view" ?

sjha4 replied:
`can_edit `and `can_view` are more record specific so they live in the permissions of individual records whereas for create, especially with empty results when there are no ACSs, we still want to know if user has create permissions for ACS in general to show/hide the create button on the UI.

ianballou reviewed (commented):
Just one comment, otherwise looks good!

ianballou commented on app/views/katello/api/v2/alternate_content_sources/permissions.rabl:
```diff
@@ -0,0 +1,8 @@
+collection @alternate_content_sources
+
+if params[:include_permissions]
```
The params here might be strings, which will always be truthy.

I curled with `/katello/api/v2/alternate_content_sources?include_permissions=false"`

ianballou commented on app/views/katello/api/v2/alternate_content_sources/base.json.rabl:
```diff
@@ -28,7 +28,7 @@ child :latest_dynflow_refresh_task => :last_refresh do |_object|
 end
 
-if params.key?(:include_permissions)
+if params[:include_permissions]
```
Same issue as the other comment.

sjha4 replied:
Updated to cast falsy values to bool.

sjha4 commented:
@ianballou : Rebased and added permissions for bulk actions. Hides bulk action when can_edit and can_delete are both false, and disable refresh/delete bulk action when one of them is false.

ianballou reviewed (commented):
For some reason, the bulk "Delete" button is staying greyed out for my user that has ACS delete permissions. The non-bulk one is not grey.

ianballou commented on app/views/katello/api/v2/alternate_content_sources/permissions.rabl:
```diff
@@ -0,0 +1,10 @@
+  node do
+    node(:can_create) { user.can?("create_alternate_content_sources") }
+    node(:can_edit) { user.can?("edit_alternate_content_sources") }
+    node(:can_delete) { user.can?("delete_alternate_content_sources") }
```
```suggestion
    node(:can_delete) { user.can?("destroy_alternate_content_sources") }
```

ianballou commented:
Should the right side kebab also be removed when the user cannot refresh nor delete?

sjha4 commented:
> Should the right side kebab also be removed when the user cannot refresh nor delete?

Updated to hide side kebab unless user has atleast edit or delete perms. And updated to refer destroy_alternate_content_sources.

ianballou commented on webpack/scenes/AlternateContentSources/MainTable/ACSTable.js:
```diff
@@ -389,9 +404,13 @@ const ACSTable = () => {
-                    <Td isActionCell>
-                      <ActionsColumn items={rowDropdownItems(acs)} />
-                    </Td>
+                    {(hasPermission(permissions, 'destroy_alternate_content_sources') ||
```
why can't you use `canDelete` and `canEdit` here?

sjha4 replied:
CanDelete and canEdit are object level. The hasPermissions will check record level permissions for the row actions which is more accurate.

ianballou reviewed (approved):
Looks good to me! I had one question, but I'm not blocking on it.
